### PR TITLE
Detach the binary tree structure from the RatchetTree

### PR DESCRIPTION
**Pull request by kkohbrok**
This PR introduces the `BinaryTree` as data structure, which is then used to represent the `RatchetTree`. With the exception of `verify_integrity`, which operates on a slice of nodes, all functions of `tree/mod.rs` have been refactored to use the `BinaryTree`. The functions provided by `BinaryTree` are somewhat tailored to what is needed by `RatchetTree`, so they're not meant to be particularly general or even useful in many other cases.

Additionally (sorry about the scope creep):
- Indexing has been improved: More (and more general) `Index` implementations, allowing us to get rid of a lot of index conversions.
- Removed the `dirpath_long` function, as it's not used anymore.
- Added some unwraps in `treemath` in places where we know that no errors will be thrown, simplifying some functions downstream.

Note, that the `BinaryTree` is not yet incorporated into the `SecretTree` (which also uses some `treemath`).

Test coverage is nothing to be excited about, especially in `tree/mod.rs`, but that should improve with the planned uptick in managed-API testing.

(Please don't look at the commit messages, they're pretty worthless. I'll squash in the end and write a proper one.)

### CODE REVIEW

File: src/tree/treemath.rs [hunk 1]
```diff
@@ -97,69 +97,52 @@ pub(crate) fn sibling(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tr
...
-        return Ok(vec![r]);
+        return vec![r];
     }
 
     let mut d = vec![];
     let mut x = index;
     while x != r {
-        x = parent(x, size)?;
+        // We can unwrap here, because we know that `index` is not the root node.
+        x = parent(x, size).unwrap();
```
Comment (raphaelrobert) [inline src/tree/treemath.rs:129]: If the index is outside of the tree this will panic.

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Great start and nice abstraction!
I like the rule of commenting on why `.unwrap()` is safe to do, we should adopt that for other PRs as well.

File: src/tree/binary_tree/mod.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        // Check if the index is within bounds to prevent `swap_remove` from
+        // panicking.
+        if node_index >= &self.size() {
+            return Err(BinaryTreeError::IndexOutOfBounds);
+        };
+        // First push the node to the end of the nodes array.
+        self.nodes.push(node);
+        // Then use `swap_remove`, which replaces the target node with the one
+        // at the end of the vector.
+        Ok(self.nodes.swap_remove(node_index.as_usize()))
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply src/tree/binary_tree/mod.rs:53]: I thought the rust-doc specifically said what is going on in `swap_remove`, i.e. it replaces the target element with the last element in the vector. So it doesn't technically preserve ordering, but in our case, it should be ok.

**Comment by kkohbrok** [comment]
@raphaelrobert Thanks for all the great feedback! I talked to @franziskuskiefer yesterday to figure out if this is indeed the abstraction we want and in the end, we were both not too certain. We agreed to put this PR on hold and reconvene for a discussion on Monday. I'll still tackle all your comments, but probably not until then, as I might have to re-write/move chunks of this code anyway.

To summarize our discussion, Franziskus thought that the implementation was too specificly focused on an array-based binary tree. At the same time, the functions are not very generic, but rather pretty aligned with the needs of MLS.

After the discussion, I think the best solution would be to choose the functions provided by `treemath` as the binary tree interface and then write a more MLS specific second layer of abstraction which contains algorithms that handle more complex operations on the tree, that are somewhere between MLS-specific and somewhat binary-tree generic. For example, the algorithms could be blank-aware (which would allow us to pull more logic into that layer), but otherwise leave node-content specific operations to the higher layers by using closures.

**Comment by kkohbrok** [comment]
This is pretty stale at this point. Closing in favour of #323.

**Pull request closed without merging**